### ISSUE TITLE
Encountered internal error running command: TypeError: Invalid Version: android-6.0.1

### ISSUE THREAD

**Issue by binubeee** [appium/appium#9926]
Iam facing this issue after upgrading my appium server and java client.
My current version:
Appium- 1.7.2
appium-java - 5.0.4

its giving error at this line:
driver = new AndroidDriver(new URL(
"http://0.0.0.0:4723/wd/hub"), capabilities);

Capabilities:
capabilities.setCapability("platformName", "android");
capabilities.setCapability("platformVersion","6.0.1");
capabilities.setCapability("deviceName", "Android");
capabilities.setCapability("udid", deviceID);
capabilities.setCapability("appPackage", pack);
capabilities.setCapability("appActivity",act);
capabilities.setCapability("clearSystemFiles", true);

Error:
[debug] [MJSONWP] Calling AppiumDriver.createSession() with args: [{"appActivity":"xxxxxxxxxx.MainActivity","appPackage":"xxxxxxxxxx","clearSystemFiles":true,"deviceName":"Android","platformName":"Android","platformVersion":"6.0.1","udid”:”6767677”},null,{"desiredCapabilities":{"appActivity":"xxxxxxxxxx.MainActivity","appPackage":"xxxxxxxxxx","clearSystemFiles":true,"deviceName":"Android","platformName":"Android","platformVersion":"6.0.1","udid":"6767677"},"firstMatch":[{"platformName":"android"}]}]
[debug] [BaseDriver] Event 'newSessionRequested' logged at 1515192722200 (17:52:02 GMT-0500 (EST))
[Appium] Merged W3C capabilities {"desiredCapabilities":{"appActivity":"com.gm.o... into desiredCapabilities object {"appActivity":"xxxxxxxxxx.MainActivity","ap...
[Appium] Creating new AndroidDriver (v1.37.0) session
[Appium] Capabilities:
[Appium] appActivity: xxxxxxxxxx.MainActivity
[Appium] appPackage: xxxxxxxxxx
[Appium] clearSystemFiles: true
[Appium] deviceName: Android
[Appium] platformName: android
[Appium] platformVersion: 6.0.1
[Appium] udid: 6767677
[BaseDriver] Session created with session id: 56291214-5319-4148-87c6-611cc7fbc7bd
[debug] [AndroidDriver] Getting Java version
[AndroidDriver] Java version is: 1.8.0_144
[ADB] Checking whether adb is present
[debug] [AndroidDriver] Shutting down Android driver
[debug] [AndroidDriver] Called deleteSession but bootstrap wasn't active
[MJSONWP] Encountered internal error running command: TypeError: Invalid Version: android-6.0.1
at new SemVer (/usr/local/lib/node_modules/appium/node_modules/semver/semver.js:293:11)
at compare (/usr/local/lib/node_modules/appium/node_modules/semver/semver.js:566:10)
at rcompare (/usr/local/lib/node_modules/appium/node_modules/semver/semver.js:576:10)
at Array.sort (native)
at ADB.callee$0$0$ (/usr/local/lib/node_modules/appium/node_modules/appium-adb/lib/tools/system-calls.js:89:17)
at tryCatch (/usr/local/lib/node_modules/appium/node_modules/babel-runtime/regenerator/runtime.js:67:40)
at GeneratorFunctionPrototype.invoke [as _invoke] (/usr/local/lib/node_modules/appium/node_modules/babel-runtime/regenerator/runtime.js:315:22)
at GeneratorFunctionPrototype.prototype.(anonymous function) [as next] (/usr/local/lib/node_modules/appium/node_modules/babel-runtime/regenerator/runtime.js💯21)
at GeneratorFunctionPrototype.invoke (/usr/local/lib/node_modules/appium/node_modules/babel-runtime/regenerator/runtime.js:136:37)
at

Can you please tell me what am I doing wrong. This was working fine before update.
Previous versions were:
Appium- 1.7.1
Java client - 4.1.2

**Comment by KazuCocoa** [comment]
Let us know in `$ANDROID_HOME/build-tools`'s directories.

```
$ ls $ANDROID_HOME/build-tools
```

In general, we can see the following result, for example, but your log looks you have `android-6.0.1` named directory.

```
23.0.1 23.0.3 25.0.0 25.0.1 25.0.2 26.0.0 26.0.2
```

**Comment by mykola-mokhnach** [comment]
I'll add a filter to not include such folders into adb verifications

**Comment by binubeee** [comment]
Yes it does!! This is what I see in that folder.

```
$ ls $ANDROID_HOME/build-tools
23.0.2		24.0.2		25.0.0		25.0.3		android-6.0.1
```

Thank you so much for getting to the root of this. I had no idea what was going on.

**Comment by binubeee** [comment]
If I may ask, by when can we expect this changes to be available for use.

Thanks,
Binu

**Comment by jiemachina** [comment]
Tks, It's useful, rename 'android-4.4w' to "20.0.0"

**Comment by rrandoom** [comment]
6666

**Comment by chenxing890** [comment]
When will appium-desktop have a new version available?

**Comment by KazuCocoa** [comment]
Which Appium version do you use?
Anyway, you needed to rename from `android-4.4w` to `20.0.0`. @zhou-sitao

**Comment by mykola-mokhnach** [comment]
.. or upgrade to appium@beta

**Comment by zhou-sitao** [comment]
Thanks for help @KazuCocoa @mykola-mokhnach 
I am using Appium- 1.7.1 and Java client - 6.0.0 beta4.

It worked after I :
1.installed adt plug-in for eclipse
2.(when I was proceeding with step one, I ran into an error 'adb server version (32) doesn't match this client (39)' and I ) uninstalled my pc application which is used to manage the android phone

But I have no idea what is the main cause of my problem yet.

**Comment by KazuCocoa** [comment]
The error happens various version's adb servers try to launch on one machine.
It happens when one adb version server has already run on your env and another tries to launch.

I'm not sure what plugins/environments influence the behaviour on your device, but some plugins conflict to launch the adb server which Appium uses. @zhou-sitao

**Comment by DivyaMahalakshmi** [comment]
Hi, I get the below error while trying to run a simple appium script
Exception in thread "main" org.openqa.selenium.SessionNotCreatedException: An unknown server-side error occurred while processing the command. Original error: Invalid Version: tools (WARNING: The server did not provide any stacktrace information)
Command duration or timeout: 0 milliseconds


The script:
import io.appium.java_client.AppiumDriver;

import java.io.File;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.concurrent.TimeUnit;

//import org.openqa.selenium.WebDriver;
import org.openqa.selenium.remote.CapabilityType;
import org.openqa.selenium.remote.DesiredCapabilities;
import org.openqa.selenium.remote.RemoteWebDriver;

import io.appium.java_client.MobileElement;
import io.appium.java_client.android.AndroidDriver;
import io.appium.java_client.remote.AndroidMobileCapabilityType;
import io.appium.java_client.remote.MobileCapabilityType;
 
public class StartApplication {
 
		//private static AndroidDriver driver;
		public static void main(String[] args) 
		{
			
		
				//throws MalformedURLException, InterruptedException {
			RemoteWebDriver driver;
			

			//Set the Desired Capabilities
			DesiredCapabilities caps = new DesiredCapabilities();
			caps.setCapability("deviceName", "Galaxy J7 Prime");
			caps.setCapability("udid", "8c1699a5"); //Give Device ID of your mobile phone
			caps.setCapability("platformName", "Android");
			caps.setCapability("platformVersion", "5.0");
			caps.setCapability("appPackage", "com.android.vending");
			caps.setCapability("appActivity", "com.google.android.finsky.activities.MainActivity");
			caps.setCapability("noReset", "true");
			
			//Instantiate Appium Driver
			try {
				driver = new RemoteWebDriver(new URL("http://127.0.0.1:4723/wd/hub"), caps);
					driver.manage().timeouts().implicitlyWait(80, TimeUnit.SECONDS);
					try {
						Thread.sleep(10000);
					} catch (InterruptedException e) {
						// TODO Auto-generated catch block
						e.printStackTrace();
					}
					driver.quit();
			}
				
			catch (MalformedURLException e) {
				System.out.println(e.getMessage());
			
			}
			
			
		}	
		
		
}
		
 Can someone please help figuring out the issue.

**Comment by KazuCocoa** [comment]
@DivyaMahalakshmi 
Appium 1.8.0 has the fix and could you run with 1.8.0 and attach the server log?

**Comment by DivyaMahalakshmi** [comment]
Hi KazuCocoa,

I was able to run the script in Appium  Version 1.5.0 (1.5.0)

Please find below the logs


The server is running

[ADB] 1 device(s) connected
[ADB] Running '/Users/divya.mahalakshmi/Library/Android/sdk/platform-tools/adb' with args: ["-P",5037,"-s","8c1699a5","shell","dumpsys","package","io.appium.unlock"]
[ADB] Checking whether aapt is present
[ADB] The installed 'io.appium.unlock' package does not require upgrade (2 >= 2)
[ADB] Getting device platform version
[ADB] Getting connected devices...
[ADB] 1 device(s) connected
[ADB] Running '/Users/divya.mahalakshmi/Library/Android/sdk/platform-tools/adb' with args: ["-P",5037,"-s","8c1699a5","shell","getprop","ro.build.version.release"]
[ADB] Current device property 'ro.build.version.release': 5.0
[ADB] Getting connected devices...
[ADB] 1 device(s) connected
[ADB] Running '/Users/divya.mahalakshmi/Library/Android/sdk/platform-tools/adb' with args: ["-P",5037,"-s","8c1699a5","shell","wm","size"]
[ADB] Getting connected devices...
[ADB] 1 device(s) connected
[ADB] Running '/Users/divya.mahalakshmi/Library/Android/sdk/platform-tools/adb' with args: ["-P",5037,"-s","8c1699a5","shell","getprop","ro.product.model"]
[ADB] Current device property 'ro.product.model': SM-G900I
[ADB] Getting connected devices...
[ADB] 1 device(s) connected
[ADB] Running '/Users/divya.mahalakshmi/Library/Android/sdk/platform-tools/adb' with args: ["-P",5037,"-s","8c1699a5","shell","getprop","ro.product.manufacturer"]
[ADB] Current device property 'ro.product.manufacturer': samsung
[AndroidDriver] No app sent in, not parsing package/activity
[AndroidDriver] No app capability. Assuming it is already on the device
[AndroidBootstrap] Watching for bootstrap disconnect
[ADB] Forwarding system: 4724 to device: 4724
[ADB] Running '/Users/divya.mahalakshmi/Library/Android/sdk/platform-tools/adb' with args: ["-P",5037,"-s","8c1699a5","forward","tcp:4724","tcp:4724"]
[UiAutomator] Starting UiAutomator
[UiAutomator] Moving to state 'starting'
[UiAutomator] Parsing uiautomator jar
[UiAutomator] Found jar name: 'AppiumBootstrap.jar'
[ADB] Running '/Users/divya.mahalakshmi/Library/Android/sdk/platform-tools/adb' with args: ["-P",5037,"-s","8c1699a5","push","/private/var/folders/vf/73hq37117wdg0wy_4pty0lmc0000gn/T/AppTranslocation/8FEC481B-1DC3-4A02-9D47-18E89B94E869/d/Appium.app/Contents/Resources/app/node_modules/appium/node_modules/appium-android-bootstrap/bootstrap/bin/AppiumBootstrap.jar","/data/local/tmp/"]
[ADB] Attempting to kill all uiautomator processes
[ADB] Getting all processes with uiautomator
[ADB] Getting connected devices...
[ADB] 1 device(s) connected
[ADB] Running '/Users/divya.mahalakshmi/Library/Android/sdk/platform-tools/adb' with args: ["-P",5037,"-s","8c1699a5","shell","ps"]
[ADB] No uiautomator process found to kill, continuing...
[UiAutomator] Starting UIAutomator
[ADB] Creating ADB subprocess with args: ["-P",5037,"-s","8c1699a5","shell","uiautomator","runtest","AppiumBootstrap.jar","-c","io.appium.android.bootstrap.Bootstrap","-e","pkg","au.org.healthdirect.macdemoapp","-e","disableAndroidWatchers",false,"-e","acceptSslCerts",false]
[UiAutomator] Moving to state 'online'
[AndroidBootstrap] [BOOTSTRAP LOG] [debug] Registered crash watchers.
[AndroidBootstrap] Android bootstrap socket is now connected
[ADB] Getting connected devices...
[AndroidBootstrap] [BOOTSTRAP LOG] [debug] Client connected
[ADB] 1 device(s) connected
[ADB] Running '/Users/divya.mahalakshmi/Library/Android/sdk/platform-tools/adb' with args: ["-P",5037,"-s","8c1699a5","shell","dumpsys","window"]
[AndroidDriver] Screen already unlocked, doing nothing
[ADB] Device API level: 21
[ADB] Getting connected devices...
[ADB] 1 device(s) connected
[ADB] Running '/Users/divya.mahalakshmi/Library/Android/sdk/platform-tools/adb' with args: ["-P",5037,"-s","8c1699a5","shell","am","start","-W","-n","au.org.healthdirect.macdemoapp/au.org.healthdirect.macdemoapp.MainActivity","-S"]
[Appium] New AndroidDriver session created successfully, session e51f2292-aefb-4642-a5e0-edd4582396f4 added to master session list
[BaseDriver] Event 'newSessionStarted' logged at 1525658234534 (11:57:14 GMT+1000 (AEST))
[MJSONWP] Responding to client with driver.createSession() result: {"platform":"LINUX","webStorageEnabled":false,"takesScreenshot":true,"javascriptEnabled":true,"databaseEnabled":false,"networkConnectionEnabled":true,"locationContextEnabled":false,"warnings":{},"desired":{"appPackage":"au.org.healthdirect.macdemoapp","appActivity":"au.org.healthdirect.macdemoapp.MainActivity","noReset":true,"platformVersion":"5.0","udid":"8c1699a5","platformName":"Android","deviceName":"Galaxy J7 Prime"},"appPackage":"au.org.healthdirect.macdemoapp","appActivity":"au.org.healthdirect.macdemoapp.MainActivity","noReset":true,"platformVersion":"5.0","udid":"8c1699a5","platformName":"Android","deviceName":"8c1699a5","deviceUDID":"8c1699a5","deviceScreenSize":"1080x1920","deviceModel":"SM-G900I","deviceManufacturer":"samsung"}
[HTTP] <-- POST /wd/hub/session 200 8042 ms - 821 
[HTTP] --> POST /wd/hub/session/e51f2292-aefb-4642-a5e0-edd4582396f4/timeouts {"type":"implicit","ms":80000}
[MJSONWP] Calling AppiumDriver.timeouts() with args: ["implicit",80000,"e51f2292-aefb-4642-a5e0-edd4582396f4"]
[BaseDriver] Set implicit wait to 80000ms
[MJSONWP] Responding to client with driver.timeouts() result: null
[HTTP] <-- POST /wd/hub/session/e51f2292-aefb-4642-a5e0-edd4582396f4/timeouts 200 5 ms - 76 
[HTTP] --> DELETE /wd/hub/session/e51f2292-aefb-4642-a5e0-edd4582396f4 {}
[MJSONWP] Calling AppiumDriver.deleteSession() with args: ["e51f2292-aefb-4642-a5e0-edd4582396f4"]
[BaseDriver] Event 'quitSessionRequested' logged at 1525658244692 (11:57:24 GMT+1000 (AEST))
[Appium] Removing session e51f2292-aefb-4642-a5e0-edd4582396f4 from our master session list
[AndroidDriver] Shutting down Android driver
[ADB] Getting connected devices...
[ADB] 1 device(s) connected
[ADB] Running '/Users/divya.mahalakshmi/Library/Android/sdk/platform-tools/adb' with args: ["-P",5037,"-s","8c1699a5","shell","am","force-stop","au.org.healthdirect.macdemoapp"]
[ADB] Pressing the HOME button
[ADB] Getting connected devices...
[ADB] 1 device(s) connected
[ADB] Running '/Users/divya.mahalakshmi/Library/Android/sdk/platform-tools/adb' with args: ["-P",5037,"-s","8c1699a5","shell","input","keyevent",3]
[AndroidBootstrap] Sending command to android: {"cmd":"shutdown"}
[AndroidBootstrap] [BOOTSTRAP LOG] [debug] Got data from client: {"cmd":"shutdown"}
[AndroidBootstrap] [BOOTSTRAP LOG] [debug] Got command of type SHUTDOWN
[AndroidBootstrap] [BOOTSTRAP LOG] [debug] Returning result: {"status":0,"value":"OK, shutting down"}
[AndroidBootstrap] Received command result from bootstrap
[UiAutomator] Shutting down UiAutomator
[UiAutomator] Moving to state 'stopping'
[AndroidBootstrap] [BOOTSTRAP LOG] [debug] Closed client connection
[AndroidBootstrap] [UIAUTO STDOUT] INSTRUMENTATION_STATUS: numtests=1
[UiAutomator] UiAutomator shut down normally
[UiAutomator] Moving to state 'stopped'
[ADB] Attempting to kill all uiautomator processes
[ADB] Getting all processes with uiautomator
[ADB] Getting connected devices...
[ADB] 1 device(s) connected
[ADB] Running '/Users/divya.mahalakshmi/Library/Android/sdk/platform-tools/adb' with args: ["-P",5037,"-s","8c1699a5","shell","ps"]
[ADB] No uiautomator process found to kill, continuing...
[UiAutomator] Moving to state 'stopped'
[Logcat] Stopping logcat capture
[ADB] Getting connected devices...
[ADB] 1 device(s) connected
[ADB] Running '/Users/divya.mahalakshmi/Library/Android/sdk/platform-tools/adb' with args: ["-P",5037,"-s","8c1699a5","shell","am","force-stop","io.appium.unlock"]
[AndroidDriver] Not cleaning generated files. Add `clearSystemFiles` capability if wanted.
[BaseDriver] Event 'quitSessionFinished' logged at 1525658246865 (11:57:26 GMT+1000 (AEST))
[MJSONWP] Received response: null
[MJSONWP] But deleting session, so not returning
[MJSONWP] Responding to client with driver.deleteSession() result: null
[HTTP] <-- DELETE /wd/hub/session/e51f2292-aefb-4642-a5e0-edd4582396f4 200 2177 ms - 76
